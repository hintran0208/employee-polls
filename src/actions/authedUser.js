export const LOGOUT_AUTHENTICATED_USER = "LOGOUT_AUTHENTICATED_USER";
export const SET_AUTHENTICATED_USER = "SET_AUTHENTICATED_USER";

export function setAuthenticatedUser(authedUser) {
  return {
    type: SET_AUTHENTICATED_USER,
    authedUser,
  };
}

export function logoutAuthedUser() {
  return {
    type: LOGOUT_AUTHENTICATED_USER,
  };
}

export function handleLogin(username, password) {
  return (dispatch, getState) => {
    const { users } = getState();

    const user = Object.values(users).find(
      (user) => user.id === username && user.password === password
    );

    if (!!user) {
      return dispatch(setAuthenticatedUser(user));
    }
  };
}

export function handleLogout() {
  return (dispatch) => {
    return dispatch(logoutAuthedUser());
  };
}
