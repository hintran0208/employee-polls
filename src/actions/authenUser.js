const LOGOUT_AUTHENTICATED_USER = "LOGOUT_AUTHENTICATED_USER";
const SET_AUTHENTICATED_USER = "SET_AUTHENTICATED_USER";

function setAuthenticatedUser(userId) {
  return {
    type: SET_AUTHENTICATED_USER,
    userId,
  };
}

function logoutAuthenticatedUser() {
  return {
    type: LOGOUT_AUTHENTICATED_USER,
  };
}

export {
  LOGOUT_AUTHENTICATED_USER,
  SET_AUTHENTICATED_USER,
  setAuthenticatedUser,
  logoutAuthenticatedUser,
};
