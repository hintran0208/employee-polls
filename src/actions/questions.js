const ADD_NEW_QUESTION = "ADD_NEW_QUESTION";
const SAVE_ANSWER_VOTE = "SAVE_ANSWER_VOTE";
const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

const receiveQuestions = (questions) => {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
};

const saveAnswerVote = (author, qid, answer) => {
  return {
    type: SAVE_ANSWER_VOTE,
    author,
    qid,
    answer,
  };
};

const addQuestion = (question) => {
  return {
    type: ADD_NEW_QUESTION,
    question,
  };
};

export {
  ADD_NEW_QUESTION,
  RECEIVE_QUESTIONS,
  SAVE_ANSWER_VOTE,
  addQuestion,
  receiveQuestions,
  saveAnswerVote,
};
