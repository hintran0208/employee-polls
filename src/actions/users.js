export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_NEW_ANSWER_USER = "ADD_NEW_ANSWER_USER";
export const ADD_NEW_QUESTION_USER = "ADD_NEW_QUESTION_USER";

function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

function addNewAnswer(authedUser, qid, answer) {
  return {
    type: ADD_NEW_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}

function addNewQuestion({ author, id }) {
  return {
    type: ADD_NEW_QUESTION_USER,
    author,
    id,
  };
}

export { addNewQuestion, addNewAnswer, receiveUsers };
