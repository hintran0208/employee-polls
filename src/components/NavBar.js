import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { handleLogout } from "../actions/authedUser";

const NavBar = ({ dispatch, authedUserId, avatarUser }) => {
  const logout = (e) => {
    e.preventDefault();
    dispatch(handleLogout());
  };

  return (
    <nav className="bg-gray-800">
      <div className="mx-auto max-w-7xl px-2 sm:px-6 lg:px-8">
        <div className="relative flex h-16 items-center justify-between">
          <div className="flex flex-1 items-center justify-center sm:items-stretch sm:justify-start">
            <div className="hidden sm:ml-6 sm:block">
              <div className="flex space-x-4">
                <Link
                  to="/"
                  className="bg-gray-900 text-white rounded-md px-3 py-2 text-sm font-medium"
                  aria-current="page"
                >
                  Home
                </Link>

                <Link
                  to="/leaderboard"
                  className="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                >
                  Leaderboard
                </Link>

                <Link
                  to="/new"
                  className="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                >
                  New Poll
                </Link>

                <button
                  onClick={logout}
                  className="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                >
                  Logout
                </button>
              </div>
            </div>
          </div>
          <div className="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0">
            <span className="font-medium px-3 py-2 text-gray-300" data-testid="user-information">
              Hello: {authedUserId}
            </span>
            <button
              type="button"
              className="rounded-full bg-gray-800 p-1 text-gray-400 hover:text-white focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
            >
              <span className="sr-only">View notifications</span>

              <svg
                className="h-6 w-6"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth="1.5"
                stroke="currentColor"
                aria-hidden="true"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  d="M14.857 17.082a23.848 23.848 0 005.454-1.31A8.967 8.967 0 0118 9.75v-.7V9A6 6 0 006 9v.75a8.967 8.967 0 01-2.312 6.022c1.733.64 3.56 1.085 5.455 1.31m5.714 0a24.255 24.255 0 01-5.714 0m5.714 0a3 3 0 11-5.714 0"
                />
              </svg>
            </button>

            <div className="relative ml-3">
              <div>
                <button
                  type="button"
                  className="flex rounded-full bg-gray-800 text-sm focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                  id="user-menu-button"
                  aria-expanded="false"
                  aria-haspopup="true"
                >
                  <span className="sr-only">Open user menu</span>
                  <img className="h-8 w-8 rounded-full" src={avatarUser} alt="" />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </nav>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  authedUserId: authedUser.id,
  avatarUser: authedUser.avatarURL,
});

export default connect(mapStateToProps)(NavBar);
