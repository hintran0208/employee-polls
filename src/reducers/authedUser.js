import { LOGOUT_AUTHENTICATED_USER, SET_AUTHENTICATED_USER } from "../actions/authedUser";

export default function authedUser(state = null, action) {
  switch (action.type) {
    case SET_AUTHENTICATED_USER:
      return action.authedUser;
    case LOGOUT_AUTHENTICATED_USER:
      return null;
    default:
      return state;
  }
}
