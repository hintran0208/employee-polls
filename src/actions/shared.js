import { receiveUsers } from "./users";
import { receiveQuestions, saveAnswerVote, addQuestion } from "./questions";
import { fetchInitialData } from "../api/api";
import { saveQuestion, saveQuestionAnswer } from "../api/api";
import { addNewAnswer, addNewQuestion } from "./users";

function handleInitialData() {
  return (dispatch) => {
    return fetchInitialData()
      .then(({ users, questions }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
      })
      .catch((error) => {
        console.log("Cannot handle init data", error);
      });
  };
}

function handleAddVote(questionId, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestionAnswer(authedUser.id, questionId, answer)
      .then(() => {
        dispatch(addNewAnswer(authedUser.id, questionId, answer));
        dispatch(saveAnswerVote(authedUser.id, questionId, answer));
      })
      .catch((error) => {
        console.log("Cannot handle add data", error);
      });
  };
}

function handleAddNewQuestion(firstVote, secondVote) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestion(firstVote, secondVote, authedUser)
      .then((question) => {
        dispatch(addQuestion(question));
        dispatch(addNewQuestion(question));
      })
      .catch((error) => {
        console.log("Cannot handle add data", error);
      });
  };
}

export { handleInitialData, handleAddVote, handleAddNewQuestion };
