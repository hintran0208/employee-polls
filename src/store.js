import { configureStore } from "@reduxjs/toolkit";
import authedUser from "./reducers/authedUser";
import questions from "./reducers/questions";
import users from "./reducers/users";

export const store = configureStore({
  reducer: {
    authedUser,
    users,
    questions,
  },
});
