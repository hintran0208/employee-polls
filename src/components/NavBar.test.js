import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../app/store";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import NavBar from "./NavBar";
import { setAuthenticatedUser } from "../actions/authedUser";

describe("NavBar", () => {
  it("should display username of logged in user", () => {
    store.dispatch(setAuthenticatedUser({ id: "sarahedo", password: "" }));

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NavBar />
        </BrowserRouter>
      </Provider>
    );

    const userSpanElement = component.getByTestId("user-information");
    expect(userSpanElement.textContent).toBe("Hello: sarahedo");
  });
});
