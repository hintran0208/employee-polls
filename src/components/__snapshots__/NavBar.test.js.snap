// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="bg-gray-800"
      >
        <div
          class="mx-auto max-w-7xl px-2 sm:px-6 lg:px-8"
        >
          <div
            class="relative flex h-16 items-center justify-between"
          >
            <div
              class="flex flex-1 items-center justify-center sm:items-stretch sm:justify-start"
            >
              <div
                class="hidden sm:ml-6 sm:block"
              >
                <div
                  class="flex space-x-4"
                >
                  <a
                    aria-current="page"
                    class="bg-gray-900 text-white rounded-md px-3 py-2 text-sm font-medium"
                    href="/"
                  >
                    Home
                  </a>
                  <a
                    class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                    href="/leaderboard"
                  >
                    Leaderboard
                  </a>
                  <a
                    class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                    href="/new"
                  >
                    New Poll
                  </a>
                  <button
                    class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                  >
                    Logout
                  </button>
                </div>
              </div>
            </div>
            <div
              class="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0"
            >
              <button
                class="rounded-full bg-gray-800 p-1 text-gray-400 hover:text-white focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  View notifications
                </span>
                <svg
                  aria-hidden="true"
                  class="h-6 w-6"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M14.857 17.082a23.848 23.848 0 005.454-1.31A8.967 8.967 0 0118 9.75v-.7V9A6 6 0 006 9v.75a8.967 8.967 0 01-2.312 6.022c1.733.64 3.56 1.085 5.455 1.31m5.714 0a24.255 24.255 0 01-5.714 0m5.714 0a3 3 0 11-5.714 0"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </button>
              <div
                class="relative ml-3"
              >
                <div>
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="flex rounded-full bg-gray-800 text-sm focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                    id="user-menu-button"
                    type="button"
                  >
                    <span
                      class="sr-only"
                    >
                      Open user menu
                    </span>
                    <img
                      alt=""
                      class="h-8 w-8 rounded-full"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="bg-gray-800"
    >
      <div
        class="mx-auto max-w-7xl px-2 sm:px-6 lg:px-8"
      >
        <div
          class="relative flex h-16 items-center justify-between"
        >
          <div
            class="flex flex-1 items-center justify-center sm:items-stretch sm:justify-start"
          >
            <div
              class="hidden sm:ml-6 sm:block"
            >
              <div
                class="flex space-x-4"
              >
                <a
                  aria-current="page"
                  class="bg-gray-900 text-white rounded-md px-3 py-2 text-sm font-medium"
                  href="/"
                >
                  Home
                </a>
                <a
                  class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
                <a
                  class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                  href="/new"
                >
                  New Poll
                </a>
                <button
                  class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                >
                  Logout
                </button>
              </div>
            </div>
          </div>
          <div
            class="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0"
          >
            <button
              class="rounded-full bg-gray-800 p-1 text-gray-400 hover:text-white focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
              type="button"
            >
              <span
                class="sr-only"
              >
                View notifications
              </span>
              <svg
                aria-hidden="true"
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.857 17.082a23.848 23.848 0 005.454-1.31A8.967 8.967 0 0118 9.75v-.7V9A6 6 0 006 9v.75a8.967 8.967 0 01-2.312 6.022c1.733.64 3.56 1.085 5.455 1.31m5.714 0a24.255 24.255 0 01-5.714 0m5.714 0a3 3 0 11-5.714 0"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
            <div
              class="relative ml-3"
            >
              <div>
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="flex rounded-full bg-gray-800 text-sm focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                  id="user-menu-button"
                  type="button"
                >
                  <span
                    class="sr-only"
                  >
                    Open user menu
                  </span>
                  <img
                    alt=""
                    class="h-8 w-8 rounded-full"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
